### PR TITLE
feat: SJIP-577 add biospecimen request stats endpoint

### PR DESCRIPTION
**Pull request by celinepelletier**
The result will be like this 

```
[
    {
        "study_code": "DS-COG-ALL",
        "study_name": "INCLUDE: (Lupo) Genomic Analysis of Congenital Heart Defects and Acute Lymphoblastic Leukemia in Children with Down Syndrome",
        "nb_participants": 1,
        "nb_available_samples": 2,
        "nb_containers": 0
    },
    {
        "study_code": "X01-deSmith",
        "study_name": "The epidemiology of transient leukemia in newborns with Down syndrome",
        "nb_participants": 2,
        "nb_available_samples": 2,
        "nb_containers": 0
    }
]
```

### CODE REVIEW

**Comment by celinepelletier** [comment]
Build fails because of dependancies that should be updated but will be in another pull request later

File: src/reports/biospecimen-request/biospecimenRequestStats.ts [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+            wantedFields,
+        );
+
+        /** Join available biospecimens by study_code */
+        const biospecimenDatasByStudy: IBiospecimenDataByStudy[] = [];
+        for (const biospecimen of availableBiospecimens) {
+            const biospecimenForStudy = availableBiospecimens.filter(
+                ({ study }) => study.study_code === biospecimen.study.study_code,
+            );
+            if (!biospecimenDatasByStudy.find(f => f.study_code === biospecimen.study.study_code)) {
```
Comment (evans-g-crsj) [inline src/reports/biospecimen-request/biospecimenRequestStats.ts:54]: I would use `some`, but it is just a suggestion.

File: src/reports/biospecimen-request/biospecimenRequestStats.ts [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+                    nb_participants: new Set(biospecimenForStudy.map(b => b.participant_fhir_id)).size,
+                    nb_available_samples: biospecimenForStudy.length,
+                    nb_containers: new Set(biospecimenForStudy.map(b => b.container_id).filter(c => c !== undefined))
+                        .size,
+                });
+            }
+        }
+
+        res.send(biospecimenDatasByStudy);
+        es.close();
```
Comment (evans-g-crsj) [inline src/reports/biospecimen-request/biospecimenRequestStats.ts:67]: why not in a finally block?

File: src/reports/biospecimen-request/biospecimenRequestStats.ts [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+    const { sqon, projectId } = req.body;
+    const userId = req['kauth']?.grant?.access_token?.content?.sub;
+    const accessToken = req.headers.authorization;
+
+    let es = null;
+    try {
+        es =
+            ES_PWD && ES_USER
+                ? new Client({ node: ES_HOST, auth: { password: ES_PWD, username: ES_USER } })
+                : new Client({ node: ES_HOST });
```
Comment (evans-g-crsj) [inline src/reports/biospecimen-request/biospecimenRequestStats.ts:29]: seems to be duplicated across the app?

**Review by evans-g-crsj — APPROVED** [review]

File: Dockerfile
Comment (evans-g-crsj) [inline Dockerfile:1]: :+1:

File: src/reports/biospecimen-request/biospecimenRequestStats.ts [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+    const { sqon, projectId } = req.body;
+    const userId = req['kauth']?.grant?.access_token?.content?.sub;
+    const accessToken = req.headers.authorization;
+
+    const wantedFields = ['sample_id', 'study.study_code', 'study.study_name', 'participant_fhir_id', 'container_id'];
+
+    let esClient = null;
+
+    try {
+        esClient = EsInstance.getInstance();
```
Comment (evans-g-crsj) [inline src/reports/biospecimen-request/biospecimenRequestStats.ts:28]: You should declare it outside of the try block. 

one of the reason is that you use it there in the catch block
```
catch (err) {
        reportGenerationErrorHandler(err, esClient);
    }
```

File: src/reports/clinical-data/index.ts [hunk 5]
```diff
@@ -24,22 +25,21 @@ const clinicalDataReport = (esHost: string) => async (req: Request, res: Respons
     } else if (p === ProjectType.kidsFirst) {
         reportConfig = configKf;
     } else {
-        console.warn("No reportConfig found.")
+        console.warn('No reportConfig found.');
     }
 
-    let es = null;
+    let esClient = null;
```
Comment (evans-g-crsj) [inline src/reports/clinical-data/index.ts:31]: same comment as above

File: src/reports/family-clinical-data/index.ts [hunk 6]
```diff
@@ -30,17 +31,17 @@ const clinicalDataReport = (esHost: string) => async (req: Request, res: Respons
         console.warn('No reportConfig found.');
     }
 
-    let es = null;
+    let esClient = null;
```
Comment (evans-g-crsj) [inline src/reports/family-clinical-data/index.ts:34]: same comment as above

File: src/reports/file-manifest/fileManifestStats.ts [hunk 7]
```diff
@@ -21,19 +21,17 @@ const fileManifestStats = () => async (req: Request, res: Response): Promise<voi
     const userId = req['kauth']?.grant?.access_token?.content?.sub;
     const accessToken = req.headers.authorization;
 
-    let es = null;
+    const wantedFields = ['file_id', 'data_type', 'size', 'participants.participant_id'];
+
+    let esClient = null;
```
Comment (evans-g-crsj) [inline src/reports/file-manifest/fileManifestStats.ts:26]: idem as above

**Review by evans-g-crsj — APPROVED** [review]

**Pull request merged**